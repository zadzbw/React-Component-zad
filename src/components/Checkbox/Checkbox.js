/**
 * Created by zad on 17/4/11.
 */
import React, {Component, PropTypes} from 'react';
import classNames from 'classnames';

export default class Checkbox extends Component {
  render() {
    return (
      <div>
        123
      </div>
    );
  }
}
