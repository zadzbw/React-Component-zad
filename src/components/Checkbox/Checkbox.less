@import "../../style/variables";
@import "../../style/mixins/index";

@checkbox-class: zad-checkbox;
@checkbox-group-class: ~"@{checkbox-class}-group";

.@{checkbox-class} {
  position: relative;
  display: inline-block;
  line-height: 1;
  vertical-align: text-bottom; // 保证假的checkbox与name在同一水平线上

  &-wrapper {
    cursor: pointer;
    font-size: @font-size-sm;
    display: inline-block;

    &:hover {
      .@{checkbox-class}-mock {
        border-color: @primary-color;
      }
    }

    &-disabled {
      cursor: not-allowed;
      .@{checkbox-class}-name {
        color: @disabled-color;
      }

      .@{checkbox-class}-mock {
        border-color: @border-color-base !important; // 边框色
        background-color: lighten(@border-color-base, 10%) !important; // 背景色

        &:after {
          animation-name: none;
          border-color: @disabled-color; // 对应的对勾或横杠颜色
        }
      }
    }
  }

  &-input:focus + &-mock {
    border-color: @primary-color;
  }

  &-input {
    position: absolute;
    left: 0;
    top: 0;
    bottom: 0;
    right: 0;
    width: 100%;
    height: 100%;
    z-index: 1;
    opacity: 0;
  }

  // 假的checkbox
  &-mock {
    position: relative;
    top: 0;
    left: 0;
    display: block;
    width: 14px;
    height: 14px;
    border: 1px solid @border-color-base;
    border-radius: @border-radius-sm;
    background-color: #fff;
    transition: all .3s;
    z-index: 2;

    // 创建一个block，利用border画出线条
    &:after {
      transform: skew(-15deg) rotate(45deg) scale(0);
      position: absolute;
      left: 4px;
      top: 1px;
      display: table;
      width: 5px;
      height: 8px;
      border: 2px solid #fff;
      border-top: 0;
      border-left: 0;
      content: ' ';
      transition: all .1s @ease-in-back;
    }
  }

  &-name {
    padding-left: 8px;
    padding-right: 8px;
  }

  &-indeterminate {
    .@{checkbox-class}-mock:after {
      transform: scale(1);
      left: 2px;
      top: 5px;
      width: 8px;
      height: 1px;
      transition: all .2s @ease-out-back .1s;
    }
  }

  &-checked {
    .@{checkbox-class}-mock:after {
      transform: skew(-15deg) rotate(45deg) scale(1);
      transition: all .2s @ease-out-back .1s;
    }
  }

  &-indeterminate, &-checked {
    .@{checkbox-class}-mock {
      background-color: @primary-color;
      border-color: @primary-color;
    }
  }
}

.@{checkbox-group-class} {
  font-size: @font-size-sm;

  &-item {
    display: inline-block;
    &:not(:last-child) {
      margin-right: 8px;
    }
  }
}
