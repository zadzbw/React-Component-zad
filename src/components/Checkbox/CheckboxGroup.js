/**
 * Created by zad on 17/4/11.
 */
import React, {PureComponent, PropTypes} from 'react';
import classNames from 'classnames';
import _omit from 'lodash/omit';
import _remove from 'lodash/remove';
import _isString from 'lodash/isString';
import _isFunction from 'lodash/isFunction';
import Checkbox, {checkboxPrefix} from './Checkbox';
import isOneOf from '../../utils/isOneOf';

const checkboxGroupPrefix = `${checkboxPrefix}-group`;

export default class CheckboxGroup extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      value: 'value' in this.props ? this.props.value : this.props.defaultValue,
    };
  }

  static propTypes = {
    value: PropTypes.arrayOf(PropTypes.string),
    defaultValue: PropTypes.arrayOf(PropTypes.string),
    options: PropTypes.array,
    onChange: PropTypes.func,
  };

  static defaultProps = {
    defaultValue: [],
    options: [],
    onChange: () => undefined,
  };

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps) {
      this.setState({
        value: nextProps.value || [],
      });
    }
  }

  changeChecked = (v) => {
    const {value} = this.state;
    const {onChange} = this.props;
    const newValue = value.concat();
    if (isOneOf(value, v)) {
      _remove(newValue, (_v) => {
        return _v === v;
      });
    } else {
      newValue.push(v);
    }
    if (!('value' in this.props)) {
      this.setState({
        value: newValue,
      });
    }
    if (_isFunction(onChange)) {
      onChange(newValue);
    }
  };

  getOptions = () => {
    const {options} = this.props;
    return options.map((item) => {
      if (_isString(item)) {
        return {name: item, value: item, disabled: false};
      }
      return item;
    });
  };

  getCheckbox = () => {
    const options = this.getOptions();
    const {value} = this.state;
    return options.map((option, i) => {
      return React.createElement(Checkbox, {
        key: `group-check-box-${i}`,
        disabled: option.disabled,
        children: option.name,
        checked: isOneOf(value, option.value),
        wrapperClassName: `${checkboxGroupPrefix}-item`,
        onChange: this.changeChecked.bind(this, option.value),
      });
    });
  };

  render() {
    const {className} = this.props;
    const restProps = _omit(this.props, [
      'className',
      'value',
      'defaultValue',
      'options',
      'onChange',
    ]);
    const groupClass = classNames(checkboxGroupPrefix, className);
    return (
      <div className={groupClass} {...restProps}>
        {this.getCheckbox()}
      </div>
    );
  }
}
