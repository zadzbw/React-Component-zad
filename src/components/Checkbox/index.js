/**
 * Created by zad on 17/4/11.
 */
import './Checkbox.less';

import Checkbox from './Checkbox';
import CheckboxGroup from './CheckboxGroup';

Checkbox.Group = CheckboxGroup;

export default Checkbox;
