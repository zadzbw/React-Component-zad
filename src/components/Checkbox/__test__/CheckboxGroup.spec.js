/**
 * Created by zad on 17/4/11.
 */
import React from 'react';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import Checkbox from '../Checkbox';
import CheckboxGroup from '../CheckboxGroup';

describe('CheckboxGroup Test', () => {
  it('controlled component', () => {
    class Test extends React.Component {
      constructor(props) {
        super(props);
        this.state = {check_group: ['Pear', 'Orange']};
      }

      testCheckGroup = (v) => {
        this.setState({
          check_group: v,
        })
      };

      render() {
        return (
          <CheckboxGroup
            options={['Apple', 'Pear', 'Orange']}
            value={this.state.check_group}
            onChange={this.testCheckGroup}
          />
        );
      }
    }

    const wrapper = shallow(<Test/>);

    expect(wrapper.state('check_group')).toContain('Pear');
    expect(wrapper.state('check_group')).toContain('Orange');
    expect(wrapper.state('check_group')).not.toContain('Apple');
    wrapper.shallow().find(Checkbox).at(0).simulate('change'); // Apple on
    wrapper.shallow().find(Checkbox).at(2).simulate('change'); // Orange off
    expect(wrapper.state('check_group')).toContain('Apple');
    expect(wrapper.state('check_group')).toContain('Pear');
    expect(wrapper.state('check_group')).not.toContain('Orange');

    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  });

  it('uncontrolled component', () => {
    const wrapper = shallow(
      <CheckboxGroup
        options={['Apple', 'Pear', 'Orange']}
        defaultValue={['Pear', 'Orange']}
      />
    );

    expect(wrapper.state().value).toContain('Pear');
    expect(wrapper.state().value).toContain('Orange');
    expect(wrapper.state().value).not.toContain('Apple');
    wrapper.find(Checkbox).at(0).simulate('change'); // Apple on
    wrapper.find(Checkbox).at(1).simulate('change'); // Pear off
    expect(wrapper.state().value).toContain('Apple');
    expect(wrapper.state().value).toContain('Orange');
    expect(wrapper.state().value).not.toContain('Pear');

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('with disabled', () => {
    const wrapper = shallow(
      <CheckboxGroup
        options={[
          {name: 'Apple', value: 'Apple'},
          {name: 'Pear', value: 'Pear'},
          {name: 'Orange', value: 'Orange', disabled: true},
        ]}
        defaultValue={['Pear', 'Orange']}
      />
    );

    expect(wrapper.find(Checkbox).at(0).prop('disabled')).toBe(false);
    expect(wrapper.find(Checkbox).at(1).prop('disabled')).toBe(false);
    expect(wrapper.find(Checkbox).at(2).prop('disabled')).toBe(true);

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
