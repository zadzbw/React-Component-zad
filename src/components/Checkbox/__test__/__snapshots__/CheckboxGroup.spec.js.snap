exports[`CheckboxGroup Test controlled component 1`] = `
<div
  className="zad-checkbox-group">
  <Checkbox
    checked={true}
    disabled={false}
    indeterminate={false}
    onChange={[Function]}
    wrapperClassName="zad-checkbox-group-item">
    Apple
  </Checkbox>
  <Checkbox
    checked={true}
    disabled={false}
    indeterminate={false}
    onChange={[Function]}
    wrapperClassName="zad-checkbox-group-item">
    Pear
  </Checkbox>
  <Checkbox
    checked={false}
    disabled={false}
    indeterminate={false}
    onChange={[Function]}
    wrapperClassName="zad-checkbox-group-item">
    Orange
  </Checkbox>
</div>
`;

exports[`CheckboxGroup Test uncontrolled component 1`] = `
<div
  className="zad-checkbox-group">
  <Checkbox
    checked={true}
    disabled={false}
    indeterminate={false}
    onChange={[Function]}
    wrapperClassName="zad-checkbox-group-item">
    Apple
  </Checkbox>
  <Checkbox
    checked={false}
    disabled={false}
    indeterminate={false}
    onChange={[Function]}
    wrapperClassName="zad-checkbox-group-item">
    Pear
  </Checkbox>
  <Checkbox
    checked={true}
    disabled={false}
    indeterminate={false}
    onChange={[Function]}
    wrapperClassName="zad-checkbox-group-item">
    Orange
  </Checkbox>
</div>
`;

exports[`CheckboxGroup Test with disabled 1`] = `
<div
  className="zad-checkbox-group">
  <Checkbox
    checked={false}
    disabled={false}
    indeterminate={false}
    onChange={[Function]}
    wrapperClassName="zad-checkbox-group-item">
    Apple
  </Checkbox>
  <Checkbox
    checked={true}
    disabled={false}
    indeterminate={false}
    onChange={[Function]}
    wrapperClassName="zad-checkbox-group-item">
    Pear
  </Checkbox>
  <Checkbox
    checked={true}
    disabled={true}
    indeterminate={false}
    onChange={[Function]}
    wrapperClassName="zad-checkbox-group-item">
    Orange
  </Checkbox>
</div>
`;
